engine/schema/src/org/apache/cloudstack/acl/RolePermissionVO.java

```diff
@@ -0,0 +1,109 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package org.apache.cloudstack.acl;
+
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.EnumType;
+import javax.persistence.Enumerated;
+import javax.persistence.GeneratedValue;
+import javax.persistence.GenerationType;
+import javax.persistence.Id;
+import javax.persistence.Table;
+import java.util.UUID;
+
+@Entity
+@Table(name = "role_permissions")
+public class RolePermissionVO implements RolePermission {
```

Comment:
@bhaisaab 
>> I'm not sure what you mean by Role refers RolePermissions, as of now each role has a list of role permissions linked to it if that's what you're asking.
With the containment approach, if there is a need to add a permission to allow say deployVM API on every role then you need to create a permission for every such role. This will lead to data duplication. Instead if the permission is created only once and then mapped to all the roles that needs it duplication won't happen. This is what I meant as reference relationship. Check the cloudstack DB with table name ending in map t o get an idea. This will also mean changing the APIs to reflect the same.